examples/flutter_gallery/lib/demo/material/expansion_panels_demo.dart

```diff
@@ -348,10 +348,11 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
                 });
               },
               children: _demoItems.map((DemoItem<dynamic> item) {
+                final dynamic builder = item.builder;
                 return new ExpansionPanel(
                   isExpanded: item.isExpanded,
                   headerBuilder: item.headerBuilder,
-                  body: item.builder(item)
+                  body: builder(item)
```

Comment:
?

Reply:
In strong mode when you write `item.builder(...)` that actually does a type check `(item.builder as DemoItemBodyBuilder<dynamic>)(...)` (because `DemoItem<T>.builder` has type  `DemoItemBodyBuilder<T>` which has `T` in contravariant position).

This type check always fails because individual items are like `DemoItem<String>`.

Reply:
/cc @leafpetersen can you confirm that my explanation is fully correct?

Reply:
The explanation is correct, but I'm surprised the fix works.  This is an annoying corner of the covariant generic checks.  It doesn't come up often, but when it does it's really confusing.  I think it's a place where something like a "wildcard" type as @Hixie and I discussed once might work better: I think it might be less confusing to users if we just treated the type of `builder` as `DemoItem<?> -> Widget` in the first place.

As it stands though, `item.builder` has static type `DemoItem<dynamic> -> Widget`.  That means that the type system will let *any* `DemoItem` be passed to the callback.  But in fact, the callback is only prepared to accept a `DemoItem<T>`, for some *particular* `T`.  Now, in this case, since you immediately pass it `item`, you are giving it the right kind of `T`, but the static type system doesn't know that, so it tries to cast `item.builder` to `DemoItemBodyBuilder<dynamic>`, which doesn't work.  

I'm surprised the fix works as stated though, since the read check is implied even when (as above) you immediately assign it to dynamic.  This issue discusses this further:

https://github.com/dart-lang/sdk/issues/31391

I would have expected that you would have had to write this as `(item as dynamic).builder(item)`.

If the pattern is always that you call `item.builder` with `item`, then the nice way to do this would be to encapsulate this in the class, which knows that it has the right kind of item for its builder.  So you'd have a method:

```
class DemoItem<T> {
  // ...
  Widget build() => builder(this);
}
```

Reply:
> I would have expected that you would have had to write this as `(item as dynamic).builder(item)`

That's how I had it written originally, but Flutter linter complains about me using `as` so I rewrote it to something that I thought is equivalent - I actually did not test it.

This seems like an extremely confusing subtlety, because under normal conditions `T v = x` and `(x as T)` are pretty much the same thing. 

> If the pattern is always that you call item.builder with item, then the nice way to do this would be to encapsulate this in the class,

Yeah, that's an option I will go with to limit the confusion.

Reply:
The key difference is where the cast happens: you need to cast to `dynamic` (either implicitly or explicitly) *before* reading the `builder` field.

Agreed that this is an extremely confusing subtlety.  :(